File: Leetcode/278. First Bad Version.ts
/**
 * The knows API is defined in the parent class Relation.
 * isBadVersion(version: number): boolean {
 *     ...
 * };
 */

var solution = function(isBadVersion: any) {

    return function(n: number): number {
        let l: number = 1;
        let r: number = n;
        
        while(l<=r) {
            let m: number = Math.floor(r - (r-l)/2);
            if( isBadVersion(m) && (m === 1 || !isBadVersion(m-1)) ) return m;
            else if( isBadVersion(m) && isBadVersion(m-1)) r = m-1;
            else l = m+1;
        }
        
        return 1;
    };
};

File: Leetcode/C++/278. First Bad Version.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int l=1, r=n;
        
        while(l<=r) {
            int m = r - (r-l)/2;
            if(isBadVersion(m) && !isBadVersion(m-1)) return m;
            else if(isBadVersion(m) && isBadVersion(m-1)) r = m-1;
            else l = m+1;
        }
        
        return -1;
    }
};
